### DISCUSSION TITLE
eglot-code-actions (mode line or cusor?) indicator

### DISCUSSION THREAD

**Discussion by kootenpv** [joaotavora/eglot#1070] [General]
I'm very satisfied with `eglot` so far for using rust, but I'd like it if I could get a visual indicator whenever a code action is possible.

Using `lsp-mode`, it shows in the mode line that there is a code action possible (i.e. shows the first entry)

Whenever yasnippet can fire, I have a different cursor color, so if it could hook into that, that would also work for me personally...

Is there a solution?

lsp examples:

![image](https://user-images.githubusercontent.com/2746227/194053921-bc34867e-c4e0-4f3a-968e-a8e94d34b438.png)

I switched to eglot, but you could also imagine on modeline even saying "Import ..." instead of `2 candidates` (I'm not interested in helm-mode, but this was the first example I could find - it also works in lp-mode without helm)

**Comment by joaotavora** [comment]
I more or less understand your problem, but not 100% porcent.  I don't understand anything from that screenshot with the yellow arrows, for example.

I understand that you want to be notified visually that a number of LSP Code Actions are possible in the "current context", meaning the position of the point in the buffer.

Can you post an image of your actual Eglot usage (preferably **without** Helm or any non-Emacs things) and tell me where you would like this information to be present.  I think the mode-line is actually a good idea.

**Comment by kootenpv** [comment]
I actually copied some code from `eglot-code-actions` and used `sideline-mode` (quite minimalistic) to display the actions in the sideline - I like it more as indicator

![image](https://user-images.githubusercontent.com/2746227/194140248-4349f5e4-0681-491c-919b-369f034c9827.png)

Basically something like this:

```emacs-lisp
(use-package eglot)

(require 'sideline)

(defun eglot-has-code-actions (beg &optional end)
  (interactive
   (if (region-active-p) `(,(region-beginning) ,(region-end)) `(,(point) nil)))
  (unless (eglot--server-capable :codeActionProvider)
    (eglot--error "Server can't execute code actions!"))
  (let* ((server (eglot--current-server-or-lose))
         (actions
          (jsonrpc-request
           server
           :textDocument/codeAction
           (list :textDocument (eglot--TextDocumentIdentifier)
                 :range (list :start (eglot--pos-to-lsp-position beg)
                              :end (eglot--pos-to-lsp-position end))
                 :context
                 `(:diagnostics
                   [,@(cl-loop for diag in (flymake-diagnostics beg end)
                               when (cdr (assoc 'eglot-lsp-diag (eglot--diag-data diag)))
                               collect it)])))))
    (remove-if (lambda (x) (string-equal x "Implement default members")) (mapcar (lambda (x) (plist-get x :title)) actions))))


(defun rust-sideline-backend-async (command)
  "Example async backend."
  (cl-case command
    (`candidates (cons :async (lambda (callback &rest _)
                                (funcall callback (mapcar (lambda (x) (concat (substring x 0 (min (length x) 50)) "  ")) (eglot-has-code-actions (point)))))))
    (`action (lambda (candidate &rest _)   ; optional
              (message "Execute command for `%s`!" candidate)))))

(setq sideline-delay 0.2)

(setq sideline-backends-right '(rust-sideline-backend-async))
```

**Comment by kootenpv** [comment]
It would be cool if it could be part of `eglot`!

Especially since `sideline` items are clickable (so no need to manually run the command and select an item).

Just my lisp is not good enough that I could get that to work (data structures, nor my knowledge of eglot/lsp)... the `action` hook is there though

**Reply by joaotavora** [reply]
> It would be cool if it could be part of eglot!

Eglot is not going to require `sideline-mode`, because `sideline-mode` is not a part of Emacs, so your snippet can't be a part of Eglot.  
But it's ok to share this snippet.  Regarding yout snippet:

* I question the need for `use-package` 
* I think `rust-sideline-backend-async` should maybe be called `eglot-sideline-code-actions-backend`.
* I think `eglot.el` could be changed so that you don't need to duplicate a big part of `eglot-code-actions` in your `eglot-has-code-actions` helper. 

Then your snippet could be just:
```lisp
(require 'cl-lib)
(require 'sideline)

(defun eglot-sideline-code-actions-backend (command)
  "Sideline backend for Eglot code actions."
  (cl-case command
    (`candidates (cons :async (lambda (callback &rest _)
                                (funcall callback (mapcar (lambda (x) (concat (substring x 0 (min (length x) 50)) "  ")) 
(eglot-code-actions (point)))))))
    (`action (lambda (candidate &rest _)   ; optional
              (message "Execute command for `%s`!" candidate)))))
(add-to-list 'sideline-backends-right 'sideline-backend-async)
```

But is kind of besides the point.

Perhaps I wasn't clear in my request.  I need to understand where you think that **vanilla** eglot **without** fancy mode-line, helms or sideline modes **could** offer a visual indication of code actions.  This is what **vanilla** Eglot looks like at its base.  

![image](https://user-images.githubusercontent.com/387011/194142607-b778680d-f2a2-49a3-ad2d-37577cd59c1a.png)

I obtained this Eglot by running from the terminal:

```
path/to/emacs-master/emacs -Q -L . -l eglot ~/tmp/coiso.cpp -f eglot -f display-line-numbers-mode
```

Other users see Eglot like this! So if you want to request adding a feature to Eglot you have to tell me where in **that** image you would like to see the visual indication of a possible code action.  There is a single code action possible on line 9.  Would you like to see it at all times, even if the cursor is not on line 9? Where would you like to see it?

**Reply by joaotavora** [reply]
Where is this `sideline.el` library?  Can you link to its source?

**Reply by joaotavora** [reply]
> I think eglot.el could be changed so that you don't need to duplicate a big part of eglot-code-actions in your eglot-has-code-actions helper.

I've not done this in the latest master

**Reply by joaotavora** [reply]
> I've not done this in the latest master

Doh, I meant: "I've **now** done this in the latest master"

**Reply by kootenpv** [reply]
Thanks a lot for your detailed response! 

>I think eglot.el could be changed so that you don't need to duplicate a big part of eglot-code-actions in your eglot-has-code-actions helper.

I will gladly use the eglot updated source so that I can remove that part of my custom snippet

> Where is this sideline.el library? Can you link to its source?

it's referred to by lsp-ui (the ui part of lsp-mode)

https://github.com/emacs-sideline/sideline

> Doh, I meant: "I've now done this in the latest master"

awesome I'll later give it a shot

**Comment by kootenpv** [comment]
> There is a single code action possible on line 9. Would you like to see it at all times, even if the cursor is not on line 9? Where would you like to see it?

- there can literally be thousands of suggestions (blame rust), so it has to be limited to the current cursor (in rust's case)
- at first I thought it's easier to just implement on the modeline (1), but now I believe (2) the sideline is a better way to do it (lsp-mode does it too)

![image](https://user-images.githubusercontent.com/2746227/194285375-fe0a32ae-4e2a-4b77-8547-2eed081224e6.png)

**Reply by kootenpv** [reply]
perhaps it makes sense to check how `lsp-ui` handles `sideline-mode` as a dependency 

I am not sure if it is optional, and I don't know how (optional) dependencies are nicely handled in elisp

it would be great if I could achieve my code snippet purely from eglot, but of course I am already happy with cleaned up master code that allows me to not reimplement the eglot functionality

**Reply by joaotavora** [reply]
Optional dependencies are possible in Eglot. Company, markdown are examples. You can try your hand at such a thing, using sideline.el though i think I'll revisit this again soon once Eglot is inside Emacs proper

**Comment by jcs090218** [comment]
People are requesting sideline-eglot in https://github.com/emacs-sideline/sideline/issues/5. I can try to submit sideline.el to Emacs if it's needed. :)

**Comment by matklad** [comment]
A screenshot of how modline solution can looks can be found in this LSP mode PR:

https://github.com/emacs-lsp/lsp-mode/pull/1781

Two notes:

* 💡 icon is traditionally used by various editors to signify presence of code actions.
* Ideally, this shouldn’t be LSP specific. The features here is a general “given the current point, here’s a bunch of things you can do”, it’s `M-x` with context, and, ideally, any mode should be able to plug its actions into this UX.

**Comment by joaotavora** [comment]
In the latest master version pushed just now (to Emacs's upstream, not here) I offer a new `eglot-execute` API function that can be used to ask a server (which you can get with the existing `eglot-current-server`) to execute one of the items returned by the existing `eglot-code-actions`.
